tabbycat/notifications/views.py

```diff
@@ -43,6 +43,71 @@ def form_valid(self, form):
         return super().form_valid(form)
 
 
+class EmailStatusView(AdministratorMixin, TournamentMixin, VueTableTemplateView):
+    page_title = gettext_lazy("Email Status")
+    page_emoji = '📤'
+
+    tables_orientation = 'rows'
+
+    def _create_status_timeline(self, status):
+        statuses = []
+        for s in status:
+            statuses.append(_("%(status)s @ %(time)s") % {'status': s.get_event_display(), 'time': s.timestamp})
+        return statuses
+
+    def get_tables(self):
+        tables = []
+        queryset = SentMessageRecord.objects.filter(
+            notification__tournament=self.tournament).select_related('notification', 'recipient').prefetch_related('emailstatus_set')
+
+        if queryset.first() is None:
+            return []
+
+        # To avoid repeat DB hits, slice based on notification
+        notifications = BulkNotification.objects.filter(tournament=self.tournament).annotate(emails=Count('sentmessagerecord'))
+
+        index = 0
+        for n in notifications:
+            emails = queryset[index:index+n.emails]
```

Comment:
This is ugly, and would prefer to have "querysets upon querysets" to neaten it up. The main problem with this is how ballot receipts would be divided up by debate and not combined by round, which would make too many tables.

Reply:
If it matters: Slicing a queryset actually does return another queryset, so long as there's no step: https://docs.djangoproject.com/en/2.1/topics/db/queries/#limiting-querysets

I haven't read your code fully, but are you just trying to retrieve email statuses, grouped by sent message record, in turn grouped by notification? If so:
``` python
notifications = self.tournament.bulknotification_set.prefetch_related('sentmessagerecord_set__emailstatus_set')
```
(You can chain prefetches, I think.)

If you want to be sure to get the `select_related` in there, I think it's either:
``` python
from django.db.models import Prefetch
notifications = self.tournament.bulknotification_set.prefetch_related(
    Prefetch('sentmessagerecord_set__emailstatus_set', queryset=SentMessageRecord.objects.select_related('recipient'))
)
```
or I'm not sure off the top of my head if its:
``` python
notifications = self.tournament.bulknotification_set.prefetch_related(
    Prefetch('sentmessagerecord_set', queryset=SentMessageRecord.objects.select_related('recipient').prefetch_related('emailstatus_set'))
)
```

Re ballot receipts, would it make sense to relax the definition of `BulkNotification` to (somehow) permit new messages to be added to an existing notification? I don't know if this would have other problematic side effects. The main one I can think of is you'd have to add `timestamp` to `SentMessageRecord` again. But for logs, I don't think redundant information is a bad thing (at worst it just gives you more data to do forensics with).

Reply:
Actually, I guess it would break the pattern if you run a second bulk notification for the same event (e.g. if you re-send draw notifications to affected participants after a change), hmm.

Reply:
I like the idea of using the same bulk notification event for all ballot receipts and have a timestamp on each individual message.

Should I put a warning somewhere that specific events could be selected (not all) and that it may quickly use up DB rows?

Reply:
What do you mean about specific events being selected?

On database rows, great question. I'm not really sure. I guess it'd be nice to have a warning somewhere, but we'd probably want some quantitative guidance? It's also not obvious to me where to put the warning.

Reply:
Quantitative guidance? Like how many rows per email to expect (assuming all events are activated)? I'd just put the warning in the documentation. The documentation would also show how to enable the feature, so it makes the most sense there.

About breaking the pattern with ballot emails, I think we could just combine results from consecutive ballot bulk notifications with the same round. If another notification gets sent in the middle of it, it isn't bad to separate the tables, but it would be good in the base case.

Reply:
Yeah. Didn't realize its being in documentation, but that makes sense!

At the moment we're treating ballot receipts as a singular special case, right? Is that how we want to architect this? I don't have a great idea for a principle that can be applied generally with the results we want right now, but having a specific hard-coded exception for one case is normally bad practice. Maybe it's the least bad for now though, as in, maybe we should just do this now and refactor it as needed if another case comes up later.